### PR TITLE
fix: resolve AttributeError on app_domain breaking SSO authentication

### LINKED ISSUE #2873: [BUG]: Critical: SSO authentication blocked - AttributeError on app_domain breaks Azure Entra ID
### 🐞 Bug Summary
**Critical:** SSO authentication completely blocked - `AttributeError` breaks Azure Entra ID integration and all SSO providers when `APP_DOMAIN` is configured. The code incorrectly treats `settings.app_domain` (a `HttpUrl` Pydantic object) as a plain string, causing a runtime crash that terminates the entire login flow.

---

### 🧩 Affected Component
Select the area of the project impacted:

- [x] `mcpgateway` - API
- [ ] `mcpgateway` - UI (admin panel)
- [ ] `mcpgateway.wrapper` - stdio wrapper
- [ ] Federation or Transports
- [ ] CLI, Makefiles, or shell scripts
- [ ] Container setup (Docker/Podman/Compose)
- [ ] Other (explain below)

**Specific file:** `mcpgateway/routers/sso.py`, lines 182-188

---

### 🔁 Steps to Reproduce

1. Deploy MCP Gateway to Azure App Service or configure `APP_DOMAIN=https://example.com` in `.env`
2. Configure Azure Entra ID SSO provider in MCP Gateway admin panel
3. Register redirect URI in Azure Portal: `https://example.com/auth/sso/callback/entra`
4. Attempt to login via Azure Entra ID SSO
5. After successful Microsoft authentication, observe crash when Azure redirects back to callback URL

**User Experience:**
1. ✅ User clicks "Login with Azure Entra ID"
2. ✅ User is redirected to Microsoft login page
3. ✅ User successfully authenticates with Microsoft
4. ❌ **Azure redirects back to callback URL** → `_validate_redirect_uri()` is called
5. ❌ **Code crashes with `AttributeError`** when validating `redirect_uri`
6. ❌ **User sees HTTP 500 Internal Server Error** instead of being logged in
7. ❌ **Authentication completely blocked - no workaround available**

---

### 🤔 Expected Behavior
The SSO callback should:
1. Validate the `redirect_uri` against configured `app_domain` without crashing
2. Extract the hostname from the `HttpUrl` object correctly
3. Complete the Azure Entra ID authentication flow
4. Log the user in and redirect to the admin panel

---

### 📓 Logs / Error Output
```python
File "mcpgateway/routers/sso.py", line 184, in _validate_redirect_uri
    app_domain = settings.app_domain.lower()
AttributeError: 'Url' object has no attribute 'lower'
```

**Problematic Code (lines 182-188):**
```python
if hasattr(settings, "app_domain") and settings.app_domain:
    # app_domain is typically just a hostname, allow both http and https
    app_domain = settings.app_domain.lower()  # ❌ CRASHES HERE - HttpUrl has no .lower()
    if redirect_host == app_domain:
        if redirect_scheme == "https" or (redirect_scheme == "http" and app_domain in ("localhost", "127.0.0.1")):
            return True
```

**Root Cause - Type Mismatch:**
In `mcpgateway/config.py:552`, `app_domain` is declared as:
```python
app_domain: HttpUrl = Field(default=HttpUrl("http://localhost:4444"))
```

`HttpUrl` is a **Pydantic type object**, not a string. Calling `.lower()` on it raises `AttributeError`.

Additionally, the code compares `redirect_host` (hostname only, e.g., `"myapp.azurewebsites.net"`) against `app_domain` (full URL object), which is logically incorrect even if `.lower()` worked.

---

### 🧠 Environment Info

| Key | Value |
|-----|-------|
| Version or commit | `main@latest` (affects all versions with SSO support) |
| Runtime | Python 3.11+ |
| Platform / OS | Azure App Service, Azure Static Web Apps, any production deployment |
| Container | Docker, Podman, or direct deployment |
| Pydantic Version | 2.x |

**Affected Deployments:**
- ✅ **Azure App Service** using `APP_DOMAIN=https://myapp.azurewebsites.net`
- ✅ **Azure Static Web Apps** with custom domains
- ✅ **Production deployments** behind Azure Front Door or Application Gateway
- ✅ **Any environment** where `APP_DOMAIN` is set to a non-localhost value

---

#### **Why This Specifically Breaks Azure Entra ID:**
Azure Entra ID **requires strict redirect_uri validation** for security:
- The redirect URI must be registered in Azure Portal (e.g., `https://myapp.azurewebsites.net/auth/sso/callback/entra`)
- MCP Gateway must validate this URI against `app_domain` before completing authentication
- When validation crashes, Azure's auth token is **rejected** and the user cannot log in
- Error manifests as HTTP 500, blocking all Azure SSO users

#### **Security Impact:**
- This bug breaks the open redirect protection mechanism
- Even when not crashing, the comparison logic is incorrect (hostname vs full URL)
- Production SSO authentication is completely non-functional

#### **Notes:**
- The comment "app_domain is typically just a hostname" is **incorrect** - the config stores full URLs
- The `hasattr()` check is unnecessary since `app_domain` has a default value
- This makes Azure Entra ID integration **completely unusable** in production
- No workaround available - authentication is hard-blocked

### PR DESCRIPTION
**Pull request by crivetimihai**
## Summary

- **Fix `AttributeError` crash** in `_validate_redirect_uri()` where `settings.app_domain` (a Pydantic `HttpUrl` object) was incorrectly treated as a plain string — `.lower()` on `HttpUrl` raises `AttributeError: 'Url' object has no attribute 'lower'`, completely blocking all SSO providers (Azure Entra ID, Google, GitHub, etc.)
- **Fix hostname-vs-URL comparison** that would never match even without the crash — `redirect_host` (e.g. `myapp.com`) was compared against a full URL object, not a hostname
- **Fix malformed production CORS origins** in `config.py` where `f"https://{self.app_domain}"` produced double-scheme URLs like `https://https://example.com/`

All three issues stem from the same root cause: `app_domain` is typed as `HttpUrl` but code assumed it was a plain string.

### Files Changed

| File | Change |
|------|--------|
| `mcpgateway/routers/sso.py` | Extract hostname from `HttpUrl` via `urlparse(str(...)).hostname` before comparison |
| `mcpgateway/config.py` | Extract hostname before constructing production CORS origins |
| `tests/unit/mcpgateway/routers/test_sso_router.py` | Use `HttpUrl` objects in test mocks to match actual runtime type |
| `CHANGELOG.md` | Add #2873 to 1.0.0-RC1 bug fixes (68→69 bugs, 125→126 total) |
| `docs/docs/architecture/roadmap.md` | Add #2873 to RC1 completed bugs |

Closes #2873

## Test plan

- [x] All 33 SSO router unit tests pass
- [x] All 174 affected tests pass (SSO router, config, settings fields, security headers)
- [x] Verified Azure Entra ID end-to-end scenario: `APP_DOMAIN=https://myapp.azurewebsites.net` correctly validates callback `https://myapp.azurewebsites.net/auth/sso/callback/entra`
- [x] Verified production CORS origins: `https://myapp.com`, `https://app.myapp.com`, `https://admin.myapp.com` (no double-scheme)
- [x] Verified edge cases: localhost HTTP allowed, evil domains rejected, HTTP rejected for production, `None` app_domain handled, relative paths always allowed

### CODE REVIEW

File: mcpgateway/routers/sso.py [hunk 1]
```diff
@@ -180,11 +180,12 @@ def _validate_redirect_uri(redirect_uri: str, request: Request | None = None) ->
...
-        app_domain = settings.app_domain.lower()
-        if redirect_host == app_domain:
+        # app_domain is an HttpUrl - extract the hostname for comparison
+        app_domain_host = urlparse(str(settings.app_domain)).hostname or ""
+        app_domain_host = app_domain_host.lower()
+        if redirect_host == app_domain_host:
             # Only allow HTTPS in production, or HTTP for localhost
-            if redirect_scheme == "https" or (redirect_scheme == "http" and app_domain in ("localhost", "127.0.0.1")):
+            if redirect_scheme == "https" or (redirect_scheme == "http" and app_domain_host in ("localhost", "127.0.0.1")):
                 return True
```
Comment (suciu-daniel) [inline mcpgateway/routers/sso.py:189]: this could be done without nesting if statements
```
if redirect_host != app_domain_host:
    return false
```

Also second part of the second if could be extracted in a variable `isLocalhost` to make it easier to read

**Comment by crivetimihai** [comment]
## Verification against running server (localhost:8080)

Tested the fix against a live MCP Gateway instance running on `localhost:8080`:

**Server config confirmed:**
- `app_domain` type: `HttpUrl` (not string) — confirms the bug would crash without the fix
- `environment`: development
- `allowed_origins`: correctly generated dev origins (no malformed URLs)

**Test results — all passed:**

| Scenario | URI | Result | Expected |
|----------|-----|--------|----------|
| Default `HttpUrl` | `http://localhost:3000/admin` | `True` | `True` |
| Relative path | `/admin` | `True` | `True` |
| Evil domain | `https://evil.com/steal` | `False` | `False` |
| Azure Entra callback | `https://myapp.azurewebsites.net/auth/sso/callback/entra` | `True` | `True` |
| Azure admin path | `https://myapp.azurewebsites.net/admin` | `True` | `True` |
| HTTP rejected (prod) | `http://myapp.azurewebsites.net/admin` | `False` | `False` |
| Open redirect blocked | `https://evil.com/phish` | `False` | `False` |
| Production CORS origins | `https://myapp.azurewebsites.net` | No double-scheme | Correct |
| 127.0.0.1 localhost | `http://127.0.0.1:8080/cb` | `True` | `True` |

**Comment by crivetimihai** [comment]
## Follow-up: edge case test coverage (commit 2)

Added 2 new tests and expanded 1 existing test to explicitly document the redirect validation policy:

- **`test_validate_redirect_uri_rejects_http_on_production_domain`** — HTTP is rejected for non-localhost domains (open redirect protection)
- **`test_validate_redirect_uri_app_domain_matches_any_port`** — `app_domain` is intentionally a domain-level (hostname-only) match; use `ALLOWED_ORIGINS` for port-specific control
- **`test_validate_redirect_uri_app_domain_localhost_http`** (expanded) — HTTP allowed for localhost on any port (development convenience)

35 tests now pass (was 33).

> **Design note:** The `app_domain` check is hostname-only by design (pre-existing behavior, not introduced by this fix). For strict origin+port matching, `ALLOWED_ORIGINS` provides exact control. This two-tier approach is intentional.
